ocaml/xapi/xapi_pool.ml

```diff
@@ -2814,3 +2814,97 @@ let remove_from_guest_agent_config ~__context ~self ~key =
   Xapi_pool_helpers.apply_guest_agent_config ~__context
 
 let rotate_secret = Xapi_psr.start
+
+let set_repository ~__context ~value =
+  let prefix = !Xapi_globs.repository_prefix in
+  let pool = Helpers.get_pool ~__context in
+  Xapi_pool_helpers.with_pool_operation
+    ~__context
+    ~self:pool
+    ~doc:"pool.set_repository"
+    ~op:`set_repository
+    (fun () ->
+      let pool = Helpers.get_pool ~__context in
+      match Db.Pool.get_repository ~__context ~self:pool with
+      | ref when ref = value -> ()
+      | _ as ref ->
+        let open Repository in
+        if ref <> value then
+          with_reposync_lock (fun () -> cleanup ~__context ~self:ref ~prefix);
+        Db.Pool.set_repository ~__context ~self:pool ~value;
+        Db.Repository.set_hash ~__context ~self:value ~value:"";
+        Db.Repository.set_up_to_date ~__context ~self:value ~value:false)
+
+let updates_sync ~__context ~force =
+  let prefix = !Xapi_globs.repository_prefix in
+  let pool = Helpers.get_pool ~__context in
+  let open Repository in
+  let hash =
+    Xapi_pool_helpers.with_pool_operation
+      ~__context
+      ~self:pool
+      ~doc:"pool.updates_sync"
+      ~op:`updates_sync
+      (fun () ->
+        let repository = get_enabled_repository ~__context in
+        if reposync_try_lock () then (
+            match force with
+            | true ->
+              finally
+                (fun () ->
+                  cleanup ~__context ~self:repository ~prefix;
+                  sync ~__context ~self:repository ~prefix;
+                  create_pool_repository ~__context ~self:repository ~prefix;
+                  let h = with_pool_repository ~__context ~self:repository ~prefix (fun () ->
+                       try_set_available_updates ~__context ~self:repository ~prefix)
+                  in
+                  Some h)
+                (fun () -> reposync_unlock ())
+            | false -> None (* still hold the reposync lock *))
+        else raise Api_errors.(Server_error (reposync_in_progress, [])))
+  in
+  match hash with
```

Comment:
This function has some issues with operations and locking.

This match statement has a separate `with_pool_operation`, which means that in theory someone can get in between and "steal" the operation, and make this function fail halfway. It's better to have a single `with_pool_operation` for the whole function.

I also don't like how the function can return in various places, where only sometimes `reposync_unlock` is called at the end, and sometimes it is unlocked earlier. It is probably fine now, but unnecessarily hard to verify, and someone may mess it up later by accident.

Why not put the whole `None` case below inside the `false` case in line 2863, and move the `finally` outside the `match force with` statement? You can then probably just use `with_reposync_lock`?

Reply:
Is the only difference between the `force` true/false cases the `cleanup` call? Then we can get rid of several duplicated lines.